# Einführung Funktionen in Python

Mit diesen Aufgaben lernen Sie wie Funktionen in Python verwendet werden können.

## Aufgabe 1 (Predict)
Lesen Sie sich diesen Code durch und versuchen Sie herauszufinden, was in diesem Code passieren könnte.

```
def franken_euro (a):
    return a * 1.09

def euro_franken (a):
    return a * 0.92

eingabe = int(input("Bitte wählen Sie für die Umrechnung von Franken zu Euro die 1 und für die Umrechnung von Euro zu Franken die 2"))

if eingabe == 1:
    franken = int(input("Bitte geben Sie den Betrag in Franken ein: "))
    print(franken_euro(franken))
elif eingabe == 2:
    euro = int(input("Bitte geben Sie den Betrag in Euro ein: "))
    print(euro_franken(euro))
else:
    print("Ungültige Eingabe!")
```

Was würde der Code ausgeben, wenn Sie als Eingabe die folgenden Zahlen wählen:
1, 7.

Zusatzfrage: Was passiert, wenn die Zahlen 9, 2 eingegeben werden?

## Aufgabe 2 (Run)
Führen Sie das Programm nun mit den Zahlen 1, 7 aus. Sind Sie auf das korrekte Ergebnis gekommen?

In [4]:
def franken_euro (a):
    return a * 1.09

def euro_franken (a):
    return a * 0.92

eingabe = int(input("Bitte wählen Sie für die Umrechnung von Franken zu Euro die 1 und für die Umrechnung von Euro zu Franken die 2"))

if eingabe == 1:
    franken = int(input("Bitte geben Sie den Betrag in Franken ein: "))
    print(franken_euro(franken))
elif eingabe == 2:
    euro = int(input("Bitte geben Sie den Betrag in Euro ein: "))
    print(euro_franken(euro))
else:
    print("Ungültige Eingabe!")


7.630000000000001


## Aufgabe 3 (Investigate & Modify)

### Teil 1
Notional Machine Aufgabe: Diese Aufgabe wird auf einem Blatt Papier gelöst. Gehen Sie dabei nochmals genau durch den Code und notieren Sie sich, was in jedem Schritt genau passiert.

### Teil 2
Im folgenden Code wurde das return durch ein print ersetzt. Überlegen Sie sich, bevor Sie den Code ausführen, was dies auslösen könnte. Testen Sie danach, ob Ihre Annahme richtig war.

In [5]:
def franken_euro (a):
    print( a * 1.09)

def euro_franken (a):
    print( a * 0.92)

eingabe = int(input("Bitte wählen Sie für die Umrechnung von Franken zu Euro die 1 und für die Umrechnung von Euro zu Franken die 2"))

if eingabe == 1:
    franken = int(input("Bitte geben Sie den Betrag in Franken ein: "))
    print(franken_euro(franken))
elif eingabe == 2:
    euro = int(input("Bitte geben Sie den Betrag in Euro ein: "))
    print(euro_franken(euro))
else:
    print("Ungültige Eingabe!")


7.630000000000001
None


Wenn der Code mit den Eingaben 1, 7 getestet wird, erhalten Sie folgendes Resultat:

- 7.630000000000001
- None

Mit print wird nun das Resultat direkt in der Funktion angezeigt. Jedoch wird nichts mehr zurückgegeben, weshalb der Aufruf keinen Wert mehr besitzt und zusätzlich None ausgegeben wird.

### Teil 3
Für das Umrechnen von Franken und Euro wird jeweils die Zahl, die umgerechnet werden soll, an die Funktion übergeben.

Nun stellt sich die Frage, ob es auch möglich ist, eine Funktion zu erstellen, der kein Parameter/Zahl mitgegeben wird. Erstellen Sie dafür eine zusätzliche Funktion mit dem Namen eingabe_ungueltig. In dieser Funktion soll "Ungültige Eingabe!" ausgegeben werden. Ersetzen Sie nun im else-Teil die direkte Ausgabe durch den Aufruf Ihrer neuen Funktion.

In [ ]:
def franken_euro (a):
    print( a * 1.09)

def euro_franken (a):
    print( a * 0.92)

eingabe = int(input("Bitte wählen Sie für die Umrechnung von Franken zu Euro die 1 und für die Umrechnung von Euro zu Franken die 2"))

if eingabe == 1:
    franken = int(input("Bitte geben Sie den Betrag in Franken ein: "))
    franken_euro(franken)
elif eingabe == 2:
    euro = int(input("Bitte geben Sie den Betrag in Euro ein: "))
    euro_franken(euro)
else:
    ## Hier soll die neue Funktion eingefügt werden.


In [ ]:
#### Lösung ####
def franken_euro (a):
    print( a * 1.09)

def euro_franken (a):
    print( a * 0.92)

def eingabe_ungueltig():
    print("Ungueltige Eingabe!")

eingabe = int(input("Bitte wählen Sie für die Umrechnung von Franken zu Euro die 1 und für die Umrechnung von Euro zu Franken die 2"))

if eingabe == 1:
    franken = int(input("Bitte geben Sie den Betrag in Franken ein: "))
    franken_euro(franken)
elif eingabe == 2:
    euro = int(input("Bitte geben Sie den Betrag in Euro ein: "))
    euro_franken(euro)
else:
    eingabe_ungueltig()

### Teil 4
Die Bank hat entschieden, eine Gebühr einzuführen. Immer wenn Geld gewechselt wird, muss der Bank 1 Franken bzw. 1 Euro bezahlt werden.
Passen Sie den Code entsprechend an. Es muss jeweils ein zweiter Wert an die Funktion übergeben werden und diese Gebühren müssen vor der Umrechnung abgezogen werden. Ändern Sie den Code so, dass die Gebühren korrekt berücksichtigt werden.

In [6]:

def franken_euro (a):
    return a * 1.09

def euro_franken (a):
    return a * 0.92

eingabe = int(input("Bitte wählen Sie für die Umrechnung von Franken zu Euro die 1 und für die Umrechnung von Euro zu Franken die 2"))

gebuehren = 1

if eingabe == 1:
    franken = int(input("Bitte geben Sie den Betrag in Franken ein: "))
    print(franken_euro(franken))
elif eingabe == 2:
    euro = int(input("Bitte geben Sie den Betrag in Euro ein: "))
    print(euro_franken(euro))
else:
    print("Ungültige Eingabe!")

KeyboardInterrupt: Interrupted by user

In [6]:
#### Lösung ####
def franken_euro (a, gebuehren):
    return (a - gebuehren) * 1.09

def euro_franken (a, gebuehren):
    return (a - gebuehren) * 0.92

eingabe = int(input("Bitte wählen Sie für die Umrechnung von Franken zu Euro die 1 und für die Umrechnung von Euro zu Franken die 2"))

gebuehren = 1

if eingabe == 1:
    franken = int(input("Bitte geben Sie den Betrag in Franken ein: "))
    print(franken_euro(franken, gebuehren))
elif eingabe == 2:
    euro = int(input("Bitte geben Sie den Betrag in Euro ein: "))
    print(euro_franken(euro, gebuehren))
else:
    print("Ungültige Eingabe!")

KeyboardInterrupt: Interrupted by user

## Aufgabe 4 (Make)

Sie erhalten den Auftrag, einen Code zu schreiben, welcher den Rabatt für Produkte in einem kleinen Laden berechnet.

Übergeben Sie der Funktion den Originalpreis und den Rabatt. Berechnen Sie den neuen Preis und geben Sie diesen zurück. Geben Sie diesen neuen Preis anschliessend aus.

- Definieren Sie eine Funktion mit einem geeigneten Namen.
- Die Funktion hat zwei Parameter.
- Verwenden Sie ein return, um das Resultat zurückzugeben.
- Fragen Sie den Benutzer nach dem Originalpreis und dem Rabatt.
- Geben Sie das Resultat aus.

In [ ]:
### Hier kann programmiert werden

In [7]:
### Lösung ###

def berechnung_preis(original_preis, rabatt):
    return original_preis * (1 - rabatt / 100)

original_preis = int(input("Bitte geben Sie den original Preis in Franken ein: "))

rabatt_preis = int(input("Bitte geben Sie den Rabatt in Franken ein: "))

neuer_preis = berechnung_preis(original_preis, rabatt_preis)
print(neuer_preis)
